docs/proposal.md

```diff
@@ -0,0 +1,262 @@
+
+    addStamp(stampName, stampDescription, stampYear, stampImageUrl);
+});
+```
+
+To facilitate code reuse, the developer has already extracted the code to add a stamp and refresh the UI into a helper function `addStamp()`:
+
+```js
+function addStamp(stampName, stampDescription, stampYear, stampImageUrl) {
+      // Add the new stamp to the collection
+    stamps.push({
+        name: stampName,
+        description: stampDescription,
+        year: stampYear,
+        imageUrl: stampImageUrl || null
+    });
+
+    // Confirm addition and update the collection
+    document.getElementById('confirmationMessage').textContent = `Stamp "${stampName}" added successfully!`;
+    renderStamps();
+}
+```
+
+To let AI agents use this functionality, the author defines the available tools. The `agent` property on the `Window` is checked to ensure the browser supports WebMCP. If supported, the `provideContext()` method is called, passing in an array of tools with a single item, a definition for the new "Add Stamp" tool. The tool accepts as parameters the same set of fields that are present in the HTML form, since this tool and the form should be functionally equivalent.
+
+```js
+if ("agent" in window) {
+    window.agent.provideContext({
+        tools: [
+            {
+                name: "add-stamp",
+                description: "Add a new stamp to the collection",
+                inputSchema: {
+                    type: "object",
+                    properties: {
+                        name: { type: "string", description: "The name of the stamp" },
+                        description: { type: "string", description: "A brief description of the stamp" },
+                        year: { type: "number", description: "The year the stamp was issued" },
+                        imageUrl: { type: "string", description: "An optional image URL for the stamp" }
+                    },
+                    required: ["name", "description", "year"]
+                },
+                async execute({ name, description, year, imageUrl }) {
+                    // TODO
+                }
+            }
+        ]
+    });
+}
+```
+
+Now the author needs to implement the tool. The tool needs to update the stamp database, and refresh the UI to reflect the change to the database. Since the code to do this is already available in the `addStamp()` function written earlier, the tool implementation is very simple and just needs to call this helper when an "add-stamp" tool call is received. After calling the helper, the tool needs to signal completion and should also provide some sort of feedback to the client application that requested the tool call. It returns a text message indicating the stamp was added:
+
+```js
+async execute({ name, description, year, imageUrl }) {
+    addStamp(name, description, year, imageUrl);
+
+    return {
+        content: [
+            {
+                type: "text",
```

Comment:
Have y'all considered non text output? Thinking through examples this seems like it'd be very useful but I'm not sure yet how it'd look.